File: ddsl-core/src/main/java/city/fini/ddsl/ast/Ast.java
package city.fini.ddsl.ast;

import city.fini.ddsl.diagnostics.Span;
import java.util.List;
import java.util.Optional;

public final class Ast {
  private Ast() {}

  public record Node<T>(T value, Span span) {}

  public record Program(List<LetDecl> lets, List<StageDecl> stages) {}

  public record LetDecl(Node<String> name, Node<String> value) {}

  public record StageDecl(Node<String> name, Node<StageKind> kind, List<StageStmt> statements, Span span) {}

  public enum StageKind {
    BUILD,
    TEST,
    IMAGE
  }

  public sealed interface StageStmt permits BaseDecl, Workdir, User, Update, ToolsBlock, CopyPaths,
      CopyArtifact, ToolBlock, ArtifactDecl, CarryTool, RuntimeBlock, Expose {}

  public record BaseDecl(Node<BaseFamily> family, Node<Variant> variant, Node<Distro> distro, Node<String> image, Span span)
      implements StageStmt {}

  public enum BaseFamily {
    CHAINGUARD
  }

  public enum Variant {
    DEV,
    RUNTIME
  }

  public enum Distro {
    WOLFI,
    DEBIAN,
    UBUNTU,
    RHEL,
    ALPINE,
    DISTROLESS,
    WINDOWS
  }

  public record Workdir(Node<String> path) implements StageStmt {}

  public record User(Node<String> user) implements StageStmt {}

  public record Update(Span span) implements StageStmt {}

  public record ToolsBlock(List<Node<String>> buildTools, Span span) implements StageStmt {}

  public record CopyPaths(List<Node<String>> sources, Node<String> dest, Span span) implements StageStmt {}

  public record CopyArtifact(Node<String> name, Optional<Node<String>> targetName, Node<String> dest, Span span)
      implements StageStmt {}

  public record ToolBlock(Node<ToolKind> kind, ToolConfig config, Span span) implements StageStmt {}

  public enum ToolKind {
    GO,
    MAVEN,
    PYTHON,
    NODE
  }

  public sealed interface ToolConfig permits GoTool, MavenTool, PythonTool, NodeTool {}

  public record GoTool(boolean build) implements ToolConfig {}

  public record MavenTool(List<String> goals, boolean skipTests) implements ToolConfig {}

  public record PythonTool(String requirements) implements ToolConfig {}

  public record NodeTool(NodeInstall install) implements ToolConfig {}

  public enum NodeInstall {
    PROD
  }

  public record ArtifactDecl(Node<String> name, Node<String> path, Span span) implements StageStmt {}

  public record CarryTool(Node<String> name, Span span) implements StageStmt {}

  public record RuntimeBlock(Node<RuntimeKind> kind, RuntimeConfig config, Span span) implements StageStmt {}

  public enum RuntimeKind {
    BINARY,
    JAVA,
    PYTHON,
    NODE
  }

  public sealed interface RuntimeConfig permits BinaryRuntime, JavaRuntime, PythonRuntime, NodeRuntime {}

  public record BinaryRuntime(String entry) implements RuntimeConfig {}

  public record JavaRuntime(String jar) implements RuntimeConfig {}

  public record PythonRuntime(String entry) implements RuntimeConfig {}

  public record NodeRuntime(String entry) implements RuntimeConfig {}

  public record Expose(Node<Integer> port) implements StageStmt {}
}

File: src/ast.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Spanned<T> {
    pub value: T,
    pub span: Span,
}

impl<T> Spanned<T> {
    pub fn new(value: T, span: Span) -> Self {
        Self { value, span }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program {
    pub lets: Vec<LetDecl>,
    pub stages: Vec<StageDecl>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LetDecl {
    pub name: Spanned<String>,
    pub value: Spanned<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StageDecl {
    pub name: Spanned<String>,
    pub kind: Spanned<StageKind>,
    pub statements: Vec<StageStmt>,
    pub span: Span,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StageKind {
    Build,
    Test,
    Image,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum StageStmt {
    Base(BaseDecl),
    Workdir(Spanned<String>),
    User(Spanned<String>),
    Update(Span),
    Tools(ToolsBlock),
    Copy(CopyStmt),
    Tool(ToolBlock),
    Produces(ArtifactDecl),
    Carry(CarryTool),
    Runtime(RuntimeBlock),
    Expose(Spanned<u16>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BaseDecl {
    pub family: Spanned<BaseFamily>,
    pub variant: Spanned<Variant>,
    pub distro: Spanned<Distro>,
    pub image: Spanned<String>,
    pub span: Span,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BaseFamily {
    Chainguard,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Variant {
    Dev,
    Runtime,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Distro {
    Wolfi,
    Debian,
    Ubuntu,
    Rhel,
    Alpine,
    Distroless,
    Windows,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ToolsBlock {
    pub build: Vec<Spanned<String>>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CopyStmt {
    Paths {
        sources: Vec<Spanned<String>>,
        dest: Spanned<String>,
        span: Span,
    },
    Artifact {
        name: Spanned<String>,
        target_name: Option<Spanned<String>>,
        dest: Spanned<String>,
        span: Span,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ToolBlock {
    pub kind: Spanned<ToolKind>,
    pub config: ToolConfig,
    pub span: Span,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ToolKind {
    Go,
    Maven,
    Python,
    Node,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ToolConfig {
    Go {
        build: bool,
    },
    Maven {
        goals: Vec<String>,
        skip_tests: bool,
    },
    Python {
        requirements: String,
    },
    Node {
        install: NodeInstall,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NodeInstall {
    Prod,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ArtifactDecl {
    pub name: Spanned<String>,
    pub path: Spanned<String>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CarryTool {
    pub name: Spanned<String>,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RuntimeBlock {
    pub kind: Spanned<RuntimeKind>,
    pub config: RuntimeConfig,
    pub span: Span,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RuntimeKind {
    Binary,
    Java,
    Python,
    Node,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RuntimeConfig {
    Binary { entry: String },
    Java { jar: String },
    Python { entry: String },
    Node { entry: String },
}
